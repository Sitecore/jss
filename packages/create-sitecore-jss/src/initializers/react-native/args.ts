import { BaseArgs, ClientAppAnswer } from '../../common';

export type ReactNativeArgs = BaseArgs & Partial<Omit<ClientAppAnswer, 'fetchWith'>>;
