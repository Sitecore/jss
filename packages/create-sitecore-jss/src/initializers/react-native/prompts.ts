import { DistinctQuestion, QuestionCollection } from 'inquirer';
import { styleguidePrompts, StyleguideAnswer } from '../../common/prompts/styleguide';
import { clientAppPrompts, ClientAppAnswer, FetchWith } from '../../common/prompts/base';

export interface ReactNativeAnswer
  extends StyleguideAnswer,
    ClientAppAnswer,
    Omit<ClientAppAnswer, FetchWith> {}

export const prompts: QuestionCollection<ReactNativeAnswer> = [
  ...clientAppPrompts.filter((p: DistinctQuestion) => p.name !== 'fetchWith'),
  ...styleguidePrompts,
];
