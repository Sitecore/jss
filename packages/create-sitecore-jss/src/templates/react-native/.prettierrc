{
  "endOfLine": "crlf",
  "singleQuote": true,
  "trailingComma": "es5",
  "tabWidth": 2,
  "arrowParens": "always",
  "printWidth": 100
}
