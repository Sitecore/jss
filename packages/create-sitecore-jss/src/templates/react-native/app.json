{
  "name": "<%- helper.getPascalCaseName(appName) %>",
  "displayName": "<%- helper.getPascalCaseName(appName) %>"
}