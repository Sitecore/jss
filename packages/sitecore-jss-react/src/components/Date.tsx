import React from 'react';
import PropTypes from 'prop-types';

export interface DateFieldProps {
  /** The date field data. */
  [htmlAttributes: string]: unknown;
  field: {
    value?: string;
    editable?: string;
  };
  /**
   * The HTML element that will wrap the contents of the field.
   */
  tag?: string;
  /**
   * Can be used to explicitly disable inline editing.
   * If true and `field.editable` has a value, then `field.editable` will be processed and rendered as component output. If false, `field.editable` value will be ignored and not rendered.
   * @default true
   */
  editable?: boolean;
  render?: (
    date: Date | null
  ) => React.ComponentClass<unknown> | React.FunctionComponent<unknown> | React.ReactNode;
}

export const DateField: React.SFC<DateFieldProps> = ({
  field,
  tag,
  editable,
  render,
  ...otherProps
}) => {
  if (!field || (!field.editable && !field.value)) {
    return null;
  }

  let children: React.ComponentType | React.ReactNode;

  const htmlProps: {
    [htmlAttr: string]: unknown;
    children?: React.ReactNode;
  } = {
    ...otherProps,
  };

  if (field.editable && editable) {
    htmlProps.dangerouslySetInnerHTML = {
      __html: field.editable,
    };
  } else if (render) {
    children = render(field.value ? new Date(field.value) : null);
  } else {
    children = field.value;
  }

  if (tag || (field.editable && editable)) {
    return React.createElement(tag || 'span', htmlProps, children);
  } else {
    return <React.Fragment>{children}</React.Fragment>;
  }
};

DateField.propTypes = {
  field: PropTypes.shape({
    value: PropTypes.any,
    editable: PropTypes.string,
  }).isRequired,
  tag: PropTypes.string,
  editable: PropTypes.bool,
  render: PropTypes.func,
};

DateField.defaultProps = {
  editable: true,
};

DateField.displayName = 'Date';
