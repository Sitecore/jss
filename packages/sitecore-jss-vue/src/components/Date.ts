import { defineComponent, h } from 'vue';

export interface FieldShape {
  value?: string;
  editable?: string;
}

export type FormatterFunction = (date: Date | null) => any;

export interface DateProps {
  /** The date field data. */
  field: FieldShape;
  /**
   * The HTML element that will wrap the contents of the field.
   */
  tag?: string;
  /**
   * Can be used to explicitly disable inline editing.
   * If true and `field.editable` has a value, then `field.editable`
   * will be processed and rendered as component output.
   * If false, `field.editable` value will be ignored and not rendered.
   */
  editable?: boolean;
  /**
   * A function that receives the date field value as a parsed JavaScript Date object.
   * The return value will be rendered as child/children of the wrapping element of the component.
   */
  formatter?: FormatterFunction;
}

export const DateField = defineComponent({
  name: 'ScDateField',
  props: {
    field: {
      type: Object,
      required: true,
    },
    tag: {
      type: String,
      default: 'span',
    },
    editable: {
      type: Boolean,
      default: true,
    },
    formatter: {
      type: Function,
      default: undefined,
    },
  },
  render() {
    const { field, editable, formatter, tag } = this.$props;
    if (!field || (!field.editable && !field.value)) {
      return null;
    }

    // this.$data should be passed along to the
    // `createElement` function in order to retain attributes and events
    // https://v3.vuejs.org/guide/render-function.html#render-functions
    const data: { [key: string]: unknown } = { ...this.$data };

    let children;
    if (field.editable && editable) {
      data.innerHTML = field.editable;
    } else if (this.$slots && this.$slots.default) {
      // if default slot is defined, pass it the date value and use the output as the return value for the `render` function.
      // note: the output of the scoped slot _should_ be a VNode or VNode array.
      const output = this.$slots.default({ date: field.value ? new Date(field.value) : null });
      return output;
    } else if (formatter) {
      children = formatter(field.value ? new Date(field.value) : null);
    } else {
      children = field.value;
    }

    return h(tag || 'span', data, children);
  },
});
