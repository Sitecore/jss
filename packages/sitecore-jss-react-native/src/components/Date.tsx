import React from 'react';
import PropTypes from 'prop-types';
import { Text } from 'react-native'
import HtmlView from 'react-native-htmlview';

export interface DateFieldProps {
	/** The date field data. */
	field: {
		value?: string;
		editable?: string;
	};
  /**
   * Can be used to explicitly disable inline editing.
   * If true and `field.editable` has a value, then `field.editable` will be processed and rendered as component output. If false, `field.editable` value will be ignored and not rendered.
   * @default true
   */
	editable?: boolean;
	render?: (date: Date | null) => React.ComponentType<any> | React.ReactNode;
	[htmlAttributes: string]: any;
}

export const DateField: React.SFC<DateFieldProps> = ({ field, editable, render, ...otherProps }) => {
	if (!field || (!field.editable && !field.value)) {
		return null;
	}

	let children: any;

	if (render) {
		children = render(field.value ? new Date(field.value) : null);
	} else {
		children = field.value
	}

	if (field.editable && editable) {
		return <HtmlView value={children} {...otherProps} />;
	} else {
		return render 
			? children
			: <Text>{children}</Text>;
	}
};

DateField.propTypes = {
	field: PropTypes.shape({
		value: PropTypes.any,
		editable: PropTypes.string,
	}).isRequired,
	editable: PropTypes.bool,
	render: PropTypes.func,
};

DateField.defaultProps = {
	editable: true,
};

DateField.displayName = 'Date';