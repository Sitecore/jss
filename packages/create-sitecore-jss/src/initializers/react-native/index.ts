import path from 'path';
import { prompt } from 'inquirer';
import { prompts, ReactNativeAnswer } from './prompts';
import { Initializer } from '../../common/Initializer';
import { transform } from '../../common/steps';
import { ReactNativeArgs } from './args';

export default class ReactNativeInitializer implements Initializer {
  get isBase() {
    return true;
  }

  async init(args: ReactNativeArgs) {
    const defaults = args.yes
      ? {
          appName: 'sitecore-jss-react-native',
          hostName: 'sitecore-jss-react-native.dev.local',
          language: '',
        }
      : {};

    const answers = await prompt<ReactNativeAnswer>(prompts, { ...defaults, ...args });

    const mergedArgs = {
      ...args,
      ...answers,
    };

    const templatePath = path.resolve(__dirname, '../../templates/react-native');
    await transform(templatePath, mergedArgs, {
      filter: (filePath: string) =>
        !!mergedArgs.language || !filePath.endsWith('{{language}}.json'),
    });

    const response = {
      appName: answers.appName,
    };

    return response;
  }
}
