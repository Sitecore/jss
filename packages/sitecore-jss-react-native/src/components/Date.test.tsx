import React from 'react';
import renderer from 'react-test-renderer';
import { View, Text } from 'react-native';
import { DateField } from './Date';

describe('<DateField />', () => {
  test('should render nothing with missing field', () => {
    const p = {
      field: {},
    };

    const c = renderer.create(<DateField {...p} />);

    const instance = c.getInstance();

    expect(instance && instance.children).toEqual(null);
    expect(c).toMatchSnapshot();
  });

  test('should render value', () => {
    const p = {
      field: {
        value: 'xxx',
      },
    };

    const c = renderer.create(<DateField {...p} />);

    expect(c).toMatchSnapshot();
  });

  test('should render value using render function', () => {
    const render = (date: Date | null) => (
      <View>
        <Text>Test test test...</Text>
        <Text>{date ? date.toDateString() : ''}</Text>
      </View>
    );

    const p = {
      field: {
        value: '10-23-2003',
      },
      render,
    };

    const c = renderer.create(<DateField {...p} />);

    expect(c).toMatchSnapshot();
  });
});
