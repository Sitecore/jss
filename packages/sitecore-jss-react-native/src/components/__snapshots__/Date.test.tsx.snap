// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DateField /> should render nothing with missing field 1`] = `null`;

exports[`<DateField /> should render value 1`] = `
<Text>
  xxx
</Text>
`;

exports[`<DateField /> should render value using render function 1`] = `
<View>
  <Text>
    Test test test...
  </Text>
  <Text>
    Thu Oct 23 2003
  </Text>
</View>
`;
